storage/repair.go

```diff
@@ -124,23 +128,280 @@ func (r shardRepairer) Repair(
 	// Add peer metadata
 	peerIter, err := session.FetchBlocksMetadataFromPeers(namespace, shard.ID(), start, end)
 	if err != nil {
-		return repair.MetadataComparisonResult{}, err
+		return repair.Result{}, err
 	}
 	if err := metadata.AddPeerMetadata(peerIter); err != nil {
-		return repair.MetadataComparisonResult{}, err
+		return repair.Result{}, err
 	}
 
 	metadataRes := metadata.Compare()
+	result := repair.Result{DifferenceSummary: metadataRes}
 
-	r.recordFn(namespace, shard, metadataRes)
+	res, err := r.repairFn(namespace, shard, metadataRes)
+	if err != nil {
+		// even in the event of repair failure, report statistics about differences
+		r.recordFn(namespace, shard, result)
+		return result, err
+	}
 
-	return metadataRes, nil
+	result.RepairSummary = res
+	r.recordFn(namespace, shard, result)
+	return result, nil
 }
 
-func (r shardRepairer) recordDifferences(
+type hostSet map[string]repair.HostBlockMetadata
+
+type blockID struct {
+	idHash ts.Hash
+	start  time.Time
+}
+
+func newHostSet() *hostSet {
+	hs := make(hostSet, 3)
+	return &hs
+}
+
+func (h *hostSet) insert(hbm repair.HostBlockMetadata) {
+	if host := hbm.Host; host != nil {
+		(*h)[host.String()] = hbm
+	}
+}
+
+func (h *hostSet) insertOneFromSet(oh *hostSet) {
+	for _, hbm := range *oh {
+		h.insert(hbm)
+		return
+	}
+}
+
+func (h *hostSet) contains(host topology.Host) bool {
+	if host == nil {
+		return false
+	}
+	_, ok := (*h)[host.String()]
+	return ok
+}
+
+func (h *hostSet) remove(host topology.Host) {
+	delete(*h, host.String())
+}
+
+func (h *hostSet) empty() bool {
+	return len(*h) == 0
+}
+
+func (r shardRepairer) newAdminRepairBlocksRequest(
+	pendingReplicasMap map[blockID]*hostSet,
+	idMap map[ts.Hash]ts.ID,
+	initialCapacity int64,
+) []block.ReplicaMetadata {
+	repairBlocks := make([]block.ReplicaMetadata, 0, initialCapacity)
+	for blkID, hs := range pendingReplicasMap {
+		for _, h := range *hs {
+			repairBlocks = append(repairBlocks, block.ReplicaMetadata{
+				ID:   idMap[blkID.idHash],
+				Host: h.Host,
+				Metadata: block.Metadata{
+					Start:    blkID.start,
+					Checksum: h.Checksum,
+					Size:     h.Size,
+				},
+			})
+		}
+	}
+	return repairBlocks
+}
+
+// newPendingReplicasMap returns a map from (series_id, block_start) -> set of host replicas,
+// using which repairs are to be performed
+//
+// NB(prateek): In the scenario when multiple peers have identical replicas for a single block, the current
+// implementation arbitrarily picks one of the peers as the source to request the replica from. In the event
+// the replica is unavailable from the selected peer, the current implementation fails. If this becomes an issue,
```

Comment:
@prateek do you want to open an issue about this?  The peer bootstrapping has a similar issue where it will retry a fetch for a block from a node that just went down but will attempt to choose another peer (which doesn't make sense as it's only valuable because its from a peer that it wanted to stream from).